File: examples/clean-api-python/interfaces/http/handlers/product_handler.py
"""Product HTTP handlers - Interface layer"""
from flask import request, jsonify
from typing import Dict, Any
import sys
import os

# Add backbone to path
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../../../..')))

from backbone.infrastructure.logging import LoggerFactory, LogLevel
from backbone.interfaces.response_builders import ProcessResponseBuilder, QueryResponseBuilder, ErrorResponseBuilder
from examples.clean_api_python.application.use_cases.create_product import (
    CreateProductUseCase,
    CreateProductInput
)
from examples.clean_api_python.application.use_cases.get_products import (
    GetProductsUseCase,
    GetProductsInput
)


class ProductHandler:
    """HTTP handlers for products"""
    
    def __init__(self, create_use_case: CreateProductUseCase, get_use_case: GetProductsUseCase):
        self.create_use_case = create_use_case
        self.get_use_case = get_use_case
        self.logger = LoggerFactory.create_layer_logger(
            "product-api",
            "interfaces",
            "ProductHandler"
        )
    
    def create_product(self):
        """Handles product creation"""
        context = self._get_request_context()
        
        self.logger.log(
            LogLevel.INFO,
            "Handling create product request",
            context=context
        )
        
        # Parse request body
        try:
            data = request.get_json()
            input_data = CreateProductInput(
                name=data.get("name"),
                description=data.get("description"),
                price=float(data.get("price", 0)),
                category=data.get("category"),
                stock=int(data.get("stock", 0))
            )
        except Exception as e:
            self.logger.log(
                LogLevel.ERROR,
                "Invalid JSON payload",
                extra_data={
                    "error": str(e),
                    "error_code": 13001001
                },
                context=context
            )
            response = ErrorResponseBuilder.bad_request(
                "Invalid JSON payload",
                {"error": str(e)}
            )
            return jsonify(response), response["status"]
        
        # Execute use case
        try:
            output = self.create_use_case.execute(input_data, context)
        except Exception as e:
            self.logger.log(
                LogLevel.ERROR,
                "Use case failed",
                extra_data={
                    "error": str(e),
                    "error_code": 13001002
                },
                context=context
            )
            response = ErrorResponseBuilder.internal_server_error(
                "Failed to create product",
                {"error": str(e)}
            )
            return jsonify(response), response["status"]
        
        # Success response
        response = ProcessResponseBuilder.created(
            "Product created successfully",
            {"product": output.product.to_dict()}
        )
        
        self.logger.log(
            LogLevel.INFO,
            "Product created successfully",
            extra_data={"product_id": output.product.id},
            context=context
        )
        
        return jsonify(response), response["status"]
    
    def get_products(self):
        """Handles getting products with filters"""
        context = self._get_request_context()
        
        self.logger.log(
            LogLevel.INFO,
            "Handling get products request",
            context=context
        )
        
        # Parse query parameters
        input_data = self._parse_query_parameters()
        
        self.logger.log(
            LogLevel.DEBUG,
            "Query parameters parsed",
            extra_data={
                "category": input_data.category,
                "min_price": input_data.min_price,
                "max_price": input_data.max_price,
                "page": input_data.page,
                "page_size": input_data.page_size
            },
            context=context
        )
        
        # Execute use case
        try:
            output = self.get_use_case.execute(input_data, context)
        except Exception as e:
            self.logger.log(
                LogLevel.ERROR,
                "Use case failed",
                extra_data={
                    "error": str(e),
                    "error_code": 13001003
                },
                context=context
            )
            response = ErrorResponseBuilder.internal_server_error(
                "Failed to get products",
                {"error": str(e)}
            )
            return jsonify(response), response["status"]
        
        # Success response with pagination
        products_data = [p.to_dict() for p in output.products]
        response = QueryResponseBuilder.success_with_pagination(
            "Products retrieved successfully",
            products_data,
            output.page,
            output.page_size,
            output.total_count
        )
        
        self.logger.log(
            LogLevel.INFO,
            "Products retrieved successfully",
            extra_data={
                "count": len(output.products),
                "total_count": output.total_count,
                "page": output.page
            },
            context=context
        )
        
        return jsonify(response), response["status"]
    
    def _parse_query_parameters(self) -> GetProductsInput:
        """Parses query parameters for filtering"""
        return GetProductsInput(
            category=request.args.get("category"),
            min_price=float(request.args.get("min_price", 0)) or None,
            max_price=float(request.args.get("max_price", 0)) or None,
            in_stock=request.args.get("in_stock") == "true",
            active=request.args.get("active") != "false",
            name_pattern=request.args.get("name"),
            page=int(request.args.get("page", 1)),
            page_size=int(request.args.get("page_size", 10)),
            sort_by=request.args.get("sort_by", "created_at"),
            sort_order=request.args.get("sort_order", "desc")
        )
    
    def _get_request_context(self) -> Dict[str, Any]:
        """Gets request context for logging"""
        return {
            "request_id": request.headers.get("X-Request-ID", "unknown"),
            "user_id": request.headers.get("X-User-ID", "unknown"),
            "method": request.method,
            "path": request.path
        }

File: backbone-go/examples/clean-api-go/interfaces/http/handlers/product_handler.go
// Package handlers contains HTTP handlers
package handlers

import (
	"encoding/json"
	"net/http"
	"strconv"
	"strings"

	"github.com/freakjazz/backbone-go/examples/clean-api-go/application/usecases"
	"github.com/freakjazz/backbone-go/examples/clean-api-go/domain/entities"
	"github.com/freakjazz/backbone-go/examples/clean-api-go/interfaces/http/middleware"
	"github.com/freakjazz/backbone-go/infrastructure/logging"
	"github.com/freakjazz/backbone-go/interfaces/responses"
)

// ProductHandler handles all product HTTP requests.
type ProductHandler struct {
	createUseCase       *usecases.CreateProductUseCase
	getListUseCase      *usecases.GetProductsUseCase
	getByIDUseCase      *usecases.GetProductByIDUseCase
	updateUseCase       *usecases.UpdateProductUseCase
	deleteUseCase       *usecases.DeleteProductUseCase
	changeStatusUseCase *usecases.ChangeProductStatusUseCase
	logger              *logging.EnhancedLogger
}

// NewProductHandler creates a new product handler.
func NewProductHandler(
	create *usecases.CreateProductUseCase,
	getList *usecases.GetProductsUseCase,
	getByID *usecases.GetProductByIDUseCase,
	update *usecases.UpdateProductUseCase,
	del *usecases.DeleteProductUseCase,
	changeStatus *usecases.ChangeProductStatusUseCase,
	logger *logging.EnhancedLogger,
) *ProductHandler {
	return &ProductHandler{
		createUseCase:       create,
		getListUseCase:      getList,
		getByIDUseCase:      getByID,
		updateUseCase:       update,
		deleteUseCase:       del,
		changeStatusUseCase: changeStatus,
		logger:              logger,
	}
}

// ----------------------------------------------------------------------------
// POST /api/products
// ----------------------------------------------------------------------------

func (h *ProductHandler) CreateProduct(w http.ResponseWriter, r *http.Request) {
	log := h.log("CreateProduct", r)
	log.Info("Handling create product request", nil)
	opts := errOpts(r)

	var input usecases.CreateProductInput
	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
		log.ErrorWithCode(msgInvalidJSON, 13001001, map[string]interface{}{"error": err.Error()})
		resp := responses.ErrorResponseBuilder.ValidationError(msgInvalidJSON, opts)
		h.sendJSON(w, resp.StatusCode, resp)
		return
	}

	output, err := h.createUseCase.Execute(r.Context(), input)
	if err != nil {
		log.ErrorWithCode("Create use case failed", 13001002, map[string]interface{}{"error": err.Error()})
		resp := responses.ErrorResponseBuilder.ValidationError(err.Error(), opts)
		h.sendJSON(w, resp.StatusCode, resp)
		return
	}

	log.Info("Product created", map[string]interface{}{"product_id": output.Product.ID})
	h.sendJSON(w, http.StatusCreated, responses.ProcessResponseBuilder.Created(output.Product.ID))
}

// ----------------------------------------------------------------------------
// GET /api/products
// ----------------------------------------------------------------------------

func (h *ProductHandler) GetProducts(w http.ResponseWriter, r *http.Request) {
	log := h.log("GetProducts", r)
	log.Info("Handling get products request", nil)

	input := h.parseQueryParameters(r)

	output, err := h.getListUseCase.Execute(r.Context(), input)
	if err != nil {
		log.ErrorWithCode("Get list use case failed", 13001003, map[string]interface{}{"error": err.Error()})
		resp := responses.ErrorResponseBuilder.InternalServerError("", errOpts(r))
		h.sendJSON(w, resp.StatusCode, resp)
		return
	}

	items := make([]map[string]interface{}, 0, len(output.Products))
	for _, p := range output.Products {
		items = append(items, productToMap(p))
	}

	resp := responses.PaginatedResponseBuilder.Success(
		items,
		int(output.TotalCount),
		output.Page,
		output.PageSize,
		"Products retrieved successfully",
	)

	log.Info("Products retrieved", map[string]interface{}{
		"count": len(items), "total": output.TotalCount,
	})
	h.sendJSON(w, http.StatusOK, resp)
}

// ----------------------------------------------------------------------------
// GET /api/products/{id}
// ----------------------------------------------------------------------------

// GetProductByID returns the raw product object — no envelope wrapper.
func (h *ProductHandler) GetProductByID(w http.ResponseWriter, r *http.Request, id string) {
	log := h.log("GetProductByID", r)
	log.Info("Handling get product by ID", map[string]interface{}{"product_id": id})

	product, err := h.getByIDUseCase.Execute(r.Context(), id)
	if err != nil {
		resp := responses.ErrorResponseBuilder.NotFound(msgProductNotFound, errOpts(r))
		h.sendJSON(w, resp.StatusCode, resp)
		return
	}

	h.sendJSON(w, http.StatusOK, responses.SimpleObjectResponseBuilder.Found(productToMap(product)))
}

// ----------------------------------------------------------------------------
// PUT /api/products/{id}
// ----------------------------------------------------------------------------

func (h *ProductHandler) UpdateProduct(w http.ResponseWriter, r *http.Request, id string) {
	log := h.log("UpdateProduct", r)
	log.Info("Handling update product request", map[string]interface{}{"product_id": id})
	opts := errOpts(r)

	var input usecases.UpdateProductInput
	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
		resp := responses.ErrorResponseBuilder.ValidationError(msgInvalidJSON, opts)
		h.sendJSON(w, resp.StatusCode, resp)
		return
	}
	input.ID = id

	product, err := h.updateUseCase.Execute(r.Context(), input)
	if err != nil {
		if strings.Contains(err.Error(), errNotFound) {
			resp := responses.ErrorResponseBuilder.NotFound(msgProductNotFound, opts)
			h.sendJSON(w, resp.StatusCode, resp)
			return
		}
		resp := responses.ErrorResponseBuilder.ValidationError(err.Error(), opts)
		h.sendJSON(w, resp.StatusCode, resp)
		return
	}

	log.Info("Product updated", map[string]interface{}{"product_id": product.ID})
	h.sendJSON(w, http.StatusOK, responses.ProcessResponseBuilder.Updated(product.ID))
}

// ----------------------------------------------------------------------------
// DELETE /api/products/{id}
// ----------------------------------------------------------------------------

func (h *ProductHandler) DeleteProduct(w http.ResponseWriter, r *http.Request, id string) {
	log := h.log("DeleteProduct", r)
	log.Info("Handling delete product request", map[string]interface{}{"product_id": id})
	opts := errOpts(r)

	if err := h.deleteUseCase.Execute(r.Context(), id); err != nil {
		if strings.Contains(err.Error(), errNotFound) {
			resp := responses.ErrorResponseBuilder.NotFound(msgProductNotFound, opts)
			h.sendJSON(w, resp.StatusCode, resp)
			return
		}
		resp := responses.ErrorResponseBuilder.InternalServerError("", opts)
		h.sendJSON(w, resp.StatusCode, resp)
		return
	}

	log.Info("Product deleted", map[string]interface{}{"product_id": id})
	h.sendJSON(w, http.StatusOK, responses.ProcessResponseBuilder.Deleted(id))
}

// ----------------------------------------------------------------------------
// PATCH /api/products/{id}/status
// ----------------------------------------------------------------------------

func (h *ProductHandler) ChangeProductStatus(w http.ResponseWriter, r *http.Request, id string) {
	log := h.log("ChangeProductStatus", r)
	log.Info("Handling change product status", map[string]interface{}{"product_id": id})
	opts := errOpts(r)

	var body struct {
		Active bool `json:"active"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		resp := responses.ErrorResponseBuilder.ValidationError(msgInvalidJSON, opts)
		h.sendJSON(w, resp.StatusCode, resp)
		return
	}

	input := usecases.ChangeProductStatusInput{ID: id, Active: body.Active}
	if err := h.changeStatusUseCase.Execute(r.Context(), input); err != nil {
		if strings.Contains(err.Error(), errNotFound) {
			resp := responses.ErrorResponseBuilder.NotFound(msgProductNotFound, opts)
			h.sendJSON(w, resp.StatusCode, resp)
			return
		}
		resp := responses.ErrorResponseBuilder.InternalServerError("", opts)
		h.sendJSON(w, resp.StatusCode, resp)
		return
	}

	log.Info("Product status changed", map[string]interface{}{"product_id": id, "active": body.Active})
	h.sendJSON(w, http.StatusOK, responses.ProcessResponseBuilder.Updated(id))
}

// ----------------------------------------------------------------------------
// helpers
// ----------------------------------------------------------------------------

func (h *ProductHandler) log(method string, r *http.Request) *logging.EnhancedLogger {
	return h.logger.
		WithLayer("interfaces").
		WithHandler("ProductHandler").
		WithMethod(method).
		WithContext(map[string]interface{}{
			"request_id": middleware.RequestIDFromContext(r.Context()),
			"method":     r.Method,
			"path":       r.URL.Path,
		})
}

func (h *ProductHandler) sendJSON(w http.ResponseWriter, status int, data interface{}) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(data)
}

// parseQueryParameters reads the 4 generic query params:
//
//	filters   — repeated, each: "field,operator,value[,condition]"
//	            operators : eq ne gt gte lt lte contains in between is_null is_not_null
//	            conditions: and (default) | or
//	            examples  : ?filters=category,eq,Electronics,and&filters=price,gt,500
//	page      — int (default 1)
//	page_size — int (default 10)
//	sort_by   — "field:direction"  e.g. price:desc  (default created_at:desc)
func (h *ProductHandler) parseQueryParameters(r *http.Request) usecases.GetProductsInput {
	q := r.URL.Query()

	page, _ := strconv.Atoi(q.Get("page"))
	if page < 1 {
		page = 1
	}
	pageSize, _ := strconv.Atoi(q.Get("page_size"))
	if pageSize < 1 {
		pageSize = 10
	}

	return usecases.GetProductsInput{
		Filters:  q["filters"], // all values of the repeated "filters" param
		Page:     page,
		PageSize: pageSize,
		SortBy:   q.Get("sort_by"),
	}
}

// productToMap converts a Product entity to a plain map for JSON responses.
func productToMap(p *entities.Product) map[string]interface{} {
	return map[string]interface{}{
		"id":          p.ID,
		"name":        p.Name,
		"description": p.Description,
		"price":       p.Price,
		"category":    p.Category,
		"stock":       p.Stock,
		"active":      p.Active,
		"created_at":  p.CreatedAt,
		"updated_at":  p.UpdatedAt,
	}
}
